shai/src/main/scala/at/forsyte/apalache/shai/TransExplorerService.scala

```diff
@@ -40,10 +68,75 @@ class TransExplorerService(connections: Ref[Map[UUID, Conn]]) extends ZioTransEx
    * @param req
    *   the request (isomorphic to the Unit)
    */
-  def openConnection(req: ConnectRequest): ZIO[ZEnv, Status, Connection] = for {
+  def openConnection(req: ConnectRequest): Result[Connection] = for {
     id <- ZIO.effectTotal(UUID.randomUUID())
     _ <- addConnection(Conn(id))
   } yield Connection(id.toString())
 
-  private def addConnection(c: Conn): UIO[Unit] = connections.update(_ + (c.id -> c))
+  /**
+   * Parses a spec into a model
+   *
+   * This method handles the [LoadModelRequest] RPC defined by `transExplorer.proto`
+   *
+   * @param req
+   *   the request to load a model, including the root module spec and any auxiliary modules
+   */
+  def loadModel(req: LoadModelRequest): Result[LoadModelResponse] = for {
+    parseResult <- parseSpec(req.spec, req.aux)
+    result <- parseResult match {
+      case Right(module) =>
+        for {
+          _ <- updateConnection(req.conn)(_.setModel(module))
+          json = jsonOfModule(module)
+        } yield LoadModelResponse.Result.Spec(json)
+      case Left(err) =>
+        ZIO.succeed(LoadModelResponse.Result.Err(err.getMessage()))
+    }
+  } yield LoadModelResponse(result)
+
+  private def parseSpec(spec: String, aux: Seq[String]): Result[Either[Throwable, TlaModule]] =
+    // Obtain permit on the semaphore protecting access to the parser, ensuring the parser is not
+    // run by more than one thread at a time.
+    parserSemaphore.withPermit(ZIO.effectTotal(try {
```

Comment:
I would rather use [Lock#tryLock](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html#tryLock(long,%20java.util.concurrent.TimeUnit)) with a large timeout (several minutes), in case SANY gets stuck somewhere for some reason. Although I am not sure what to do if SANY really gets stuck. Perhaps hide it in a thread that is killed by a timeout?

Reply:
We can add a timeout, but not with lock try. Every effect can have a timeout added in ZIO. See https://zio.dev/version-1.x/overview/overview_basic_concurrency#timeout

Reply:
Timeout added in 49948dd7cb90696110e2d2f7136773c1e0de1fd4